src/java.base/share/classes/javax/crypto/KDF.java

```diff
@@ -0,0 +1,584 @@
+     *     algorithm or {@code null} if no additional parameters are provided
+     * @param provider
+     *     the provider to use for this key derivation; if null, this method is
+     *     equivalent to {@code getInstance(String, AlgorithmParameterSpec)}
+     *
+     * @return a {@code KDF} object
+     *
+     * @throws NoSuchAlgorithmException
+     *     if a provider is specified and it does not support the specified KDF
+     *     algorithm, or if provider is {@code null} and there is no provider
+     *     that supports a KDF implementation of the specified algorithm
+     * @throws InvalidAlgorithmParameterException
+     *     if the {@code AlgorithmParameterSpec} is an invalid value
+     * @throws NullPointerException
+     *     if the algorithm is {@code null}
+     */
+    public static KDF getInstance(String algorithm,
+                                  AlgorithmParameterSpec algParameterSpec,
+                                  Provider provider)
+        throws NoSuchAlgorithmException, InvalidAlgorithmParameterException {
+        Objects.requireNonNull(algorithm, "null algorithm name");
+        try {
+            Instance instance = GetInstance.getInstance("KDF", KDFSpi.class,
+                                                        algorithm,
+                                                        algParameterSpec,
+                                                        provider);
+            if (!JceSecurity.canUseProvider(instance.provider)) {
+                String msg = "JCE cannot authenticate the provider "
+                             + instance.provider.getName();
+                throw new SecurityException(msg);
+            }
+            return new KDF((KDFSpi) instance.impl, instance.provider, algorithm,
+                           algParameterSpec);
+
+        } catch (NoSuchAlgorithmException nsae) {
+            return handleException(nsae);
+        }
+    }
+
+    private static KDF handleException(NoSuchAlgorithmException e)
+        throws NoSuchAlgorithmException,
+               InvalidAlgorithmParameterException {
+        Throwable cause = e.getCause();
+        if (cause instanceof InvalidAlgorithmParameterException) {
+            throw (InvalidAlgorithmParameterException) cause;
+        }
+        throw e;
+    }
+
+    /**
+     * Derives a key, returned as a {@code SecretKey}.
+     * <p>
+     * The {@code deriveKey} method may be called multiple times on a particular
+     * {@code KDF} instance.
+     * <p>
+     * Delayed provider selection is also supported such that the provider
+     * performing the derive is not selected until the method is called. Once a
+     * provider is selected, it cannot be changed.
+     *
+     * @param alg
+     *     the algorithm of the resultant {@code SecretKey} object (may not be
+     *     {@code null})
+     * @param kdfParameterSpec
+     *     derivation parameters (may not be {@code null}
```

Comment:
Normally, we only say if parameters may be null and don't say parameters may not be null, since the `@throws NullPointerException` makes this case clear. So I would remove the words "(may not be null)".

Same comment for deriveData.